### DISCUSSION TITLE
What is the difference between IDispatcher and ICommandProcessor?

### DISCUSSION THREAD

**Discussion by mehdihadeli** [BrighterCommand/Brighter#1469] [Q&A]
Hi @iancooper,
What is your purpose for defining these two separate classes `Dispatcher` and `CommandProcessor`,  for dispatching mechanism?
I think `CommandProcessor` itself is enough!
In my case I don't want to use message brokers and I want to dispatch command to a command handler.

**Comment by iancooper** [comment]
Hi @mehdihadeli,

The CommandProcessor take your request and routes it to one or more handlers. On a Send or Publish these are in process. On a Call or Post these are out-of-process and go over middleware (RMQ, Kafka, ASB etc.)

The Dispatcher lets you listen for a message on middleware (RMQ, Kafka, ASB etc) and routes it, using the CommandHandler, to a handler.

So Dispatcher runs a message pump that listens for incoming messages in the same way that ASP.NET listens for HTTP messages and routes them to a Controller.

If you don't want to use middleware, you just need the CommandProcessor

**Reply by mehdihadeli** [reply]
@iancooper Thanks for your Help :)
So `Dispatcher` class is available only for out-of-process commands?
